source/common/router/rds_impl.h

```diff
@@ -138,21 +122,62 @@ class RdsRouteConfigProviderImpl
   void registerInitTarget(Init::Manager& init_manager);
   void runInitializeCallbackIfAny();
 
-  Server::Configuration::FactoryContext& factory_context_;
   std::unique_ptr<Envoy::Config::Subscription<envoy::api::v2::RouteConfiguration>> subscription_;
-  ThreadLocal::SlotPtr tls_;
   std::string config_source_;
+  std::function<void()> initialize_callback_;
   const std::string route_config_name_;
-  absl::optional<LastConfigInfo> config_info_;
   Stats::ScopePtr scope_;
   RdsStats stats_;
-  std::function<void()> initialize_callback_;
   RouteConfigProviderManagerImpl& route_config_provider_manager_;
   const std::string manager_identifier_;
-  envoy::api::v2::RouteConfiguration route_config_proto_;
+  SystemTimeSource& time_source_;
   SystemTime last_updated_;
+  absl::optional<LastConfigInfo> config_info_;
+  envoy::api::v2::RouteConfiguration route_config_proto_;
+  std::unordered_set<RdsRouteConfigProviderImpl*> route_config_providers_;
+
+  friend class RouteConfigProviderManagerImpl;
+  friend class RdsRouteConfigProviderImpl;
+};
+
+typedef std::shared_ptr<RdsRouteConfigSubscription> RdsRouteConfigSubscriptionSharedPtr;
+
+/**
+ * Implementation of RouteConfigProvider that fetches the route configuration dynamically using
+ * the subscription.
+ */
+class RdsRouteConfigProviderImpl : public RouteConfigProvider,
+                                   public std::enable_shared_from_this<RdsRouteConfigProviderImpl>,
+                                   Logger::Loggable<Logger::Id::router> {
+public:
+  ~RdsRouteConfigProviderImpl();
+
+  RdsRouteConfigSubscription& subscription() { return *subscription_; }
+
+  // Router::RouteConfigProvider
+  Router::ConfigConstSharedPtr config() override;
+  absl::optional<ConfigInfo> configInfo() const override;
+  SystemTime lastUpdated() const override { return subscription_->last_updated_; }
+
+private:
+  struct ThreadLocalConfig : public ThreadLocal::ThreadLocalObject {
+    ThreadLocalConfig(ConfigConstSharedPtr initial_config) : config_(initial_config) {}
+
+    ConfigConstSharedPtr config_;
+  };
+
+  RdsRouteConfigProviderImpl(RdsRouteConfigSubscriptionSharedPtr&& subscription,
+                             Server::Configuration::FactoryContext& factory_context);
+
+  void onConfigUpdate();
+
+  RdsRouteConfigSubscriptionSharedPtr subscription_;
+  Server::Configuration::FactoryContext& factory_context_;
+  ThreadLocal::SlotPtr tls_;
+  const std::string route_config_name_;
 
   friend class RouteConfigProviderManagerImpl;
+  friend class RdsRouteConfigSubscription;
```

Comment:
Why does Subscription need to access ProviderImpl private members?

Reply:
mostly for for onConfigUpdate since no class other than Subscription should call it. will make it public and remove friend here.

Reply:
If we use SubscriptionCallback interface,  this friend class will not be needed.